Guides/organizing dbt cloud jobs (#3672)
dbt-labs/docs.getdbt.com: website/docs/guides/orchestration/running-dbt-cloud-jobs/1-types-of-jobs.md

---
title: "Organizing dbt Cloud jobs"
id: "organizing-dbt-cloud-jobs"
slug: 1-types-of-jobs
description: Learn how to organize your dbt Cloud jobs by category
displayText: The types of jobs in dbt Cloud
hoverSnippet: Learn how to organize your dbt Cloud jobs by category.
---

dbt Cloud jobs perform two main functions:

- Running models in production
- Performing Continuous Integration Testing on models during development

> dbt Cloud jobs can optionally run source freshness checks generate documentation as well.

Reviewer: ```suggestion
dbt Cloud jobs can optionally run source freshness checks and generate documentation as well.
```

<callout>

> Make sure to leverage the dbt build command to avoid wasteful spend! dbt build enables you to run tests for each individual model, before it's downstream dependencies start. When a test fails, the dependences of the model that failed testing are skipped. This enables you to avoid wasted spend on downstream node runs when an upstream node fails a test.

Reviewer: ```suggestion
Make sure to leverage the dbt build command to avoid wasteful spend! dbt build enables you to run tests for each individual model, before its downstream dependencies start. When a test on a model fails, the dependences of that model are skipped. This enables you to avoid wasted spend on downstream node runs.
```

</callout>

### Running Models in Production

#### Organizing models using tags

> To better organize your models when developing the orchestration strategy, it's useful to classify the models in your DAG along dimensions such as:

Reviewer: ```suggestion
Organizing your models in groups will allow you to create a successful orchestration strategy. Common classifications that users employ are:
```

> - Frequency: do they need to run Hourly, Daily, Weekly

Reviewer: ```suggestion
- Frequency: hourly, daily, weekly build schedules
```

- Sensitivity: PII, GDPR, HIPAA
- Domain: Marketing, Finance, Ops

> - Readiness: Staging, Intermediate, Marts / Bronze, Silver, Gold

Reviewer: would you have all of these? Should you just say staging, intermediate, marts? I thought Gold = Marts?

Author: Yeah good question. I was trying to catch both the "Medallion" audience and others in the same concept. Will figure out better phrasing.

Reviewer 2: Maybe you could either have them as two bullets, or have the bullet called Readiness/Medallion Level?

> - dbt Configuration: Tables, Views, Incrementals, Snapshots

Reviewer 3: Is this needed? Because it would be easier and more efficient to use model.materialisation rather than tags for scheduling purposes

Reviewer 2: ~Agree, this seems weird~ just saw further down the page where you talk about incrementals and snapshots specifically - I could get on board with that strategy, but calling out views and tables doesn't make sense, assuming you select them by materialization instead of tags

<callout>

> Frequency is most commonly used to determine run cadence. A good strategy is to tag models at the "end" of the DAG with the update frequency you require (e.g. hourly, daily). Then structure the dbt job command as `dbt build -s +tag:hourly` in order to automatically identify the models upstream of the tagged models which also need to run.

Reviewer: ```suggestion
Frequency is most commonly used to determine run cadence. A good strategy is to tag models at the "end" of the DAG with the build frequency you require (e.g. hourly, daily). Then structure the dbt job command as `dbt build -s +tag:hourly` in order to automatically select the models upstream of, and including, the tagged models.
```

</callout>

#### Running tagged models

##### Production Hourly

`dbt build -s +tag:hourly`

##### Production Daily

`dbt build -s +tag:daily`

##### Managing PII & Masking Policies

> Organizations that deal with PII data generally leverage data masking policies to protect it from unintended use. [reference dbt snowmask package]. In this situation, it makes sense to separate the builds include PII data masking policies, in particular because doing so makes it much easier to manage logging and auditing.

Reviewer: ```suggestion
Organizations that work with PII data generally leverage data masking policies to protect sensitive data from unintended use. [reference dbt snowmask package]. In this situation, we recommend you separate the builds including PII data masking policies from other jobs, as additional isolation makes logging and auditing attribution easier.
```

Reviewer 2: Looks like a missed todo: "reference dbt snowmask pkg"

`dbt build -s +tag:pii` # builds static data masks
`dbt build -s tag:hourly --exclude tag:pii` # build hourly data downstream of masks

You can also do things like `dbt build -s tag:every_four_hours` for more fine-grained scheduling control.

### Running Models by dbt Configuration

> Another approach that works well is to segment dbt Cloud jobs according to model configuration type. Of the items below, `incremental` and `snapshot` models are often the ones which make the most sense to separate from the others onto their own run cadences. 

Reviewer: ```suggestion
Another approach that works well is to segment dbt Cloud jobs according to model configuration type. `Incremental` and `Snapshot` models are often separated into their own job runs. 
```

Reviewer 2: As a rule, we won't code-format words when they're not in a code context, you can just say Incremental and snapshot models.

> For `incremental` models, performance is often a concern, which leads dbt Cloud users to create distinct run cadences for that subset. You can also create a second dbt Cloud job to run `incremental` models with the `--full-refresh` flag included to enable periodic full rebuilds.

Reviewer: ```suggestion
Users employ an `incremental` materialization strategy when concerned about model cost and performance. Often, users will create a distinct, frequent job run cadence to build models incrementally throughout the week, achieving a cost-effective way to get up-to-date data regularly. Users will also have a separate job to build these incremental models from scratch on a less frequent basis. We recommend that users set up a job to build `incremental` models with the `--full-refresh` flag to run at least weekly to ensure data quality and completeness.
```

> dbt Labs recommends treating `snapshot` data differently than the rest of the DAG. A good approach is to create a separate dbt Cloud Environment including a different permission set to run snapshots and materialize them into a different schema than the rest of the models. These extra precautions are beneficial because snapshot data is stateful, i.e. can be expensive to recreate if the history is lost! 

Reviewer: why separate environment and not just a schema override?

Reviewer 2: Agreed! We normally recommend building into a separate schema, ideally one where users don't have permission to drop objects. 

"can be expensive to recreate" normally it's impossible to recreate - if you could recreate it some other way, you should probably do that instead of snapshotting.

##### Tables

`dbt build -s config.materialized:table`

##### Views

> `dbt build -s config.materialized:view`

Reviewer 2: When would you ever do this?

##### Snapshots

> `dbt build -s config.materialized:snapshot`

Reviewer 2: At this point, surely just `dbt snapshot`?

##### Incrementals

`dbt build -s config.materialized:incremental`

<callout>

> NOTE: Having `incremental` models broken into a separate job is also beneficial in the event you need to `--full-refresh` them!

Reviewer: ```suggestion
NOTE: Having `incremental` models broken into a separate job is also beneficial in the event you need to `--full-refresh` them.
```

</callout>

### Continuous Integration Testing

There are two main operations performed for Continuous Integration Testing in dbt Cloud:

- Compiling the codebase 

> - Running dbt tests

Reviewer 2: I disagree with these - I talk about compilation below, and "Running dbt tests" feels too reductive. It's building the models and then running tests against them, not just the tests alone

#### Compiling the codebase

Set up a dbt Cloud job in the QA environment which runs the `dbt compile` command only

> Running this job about once an hour typically works well. Increase the frequency for very active codebases. 

Reviewer 2: To achieve what? I assume you're talking about creating artifacts that can be deferred against? There's no benefit to explicitly compiling as part of a CI flow, since this will happen on every commit anyway

#### Running Tests

Tests are run in a Continuous integration test job which defers to the compilation job. With this setup, you are able to automatically identify *only models whose code changed* in the pull request and run just those models. This keeps test runs fast and 
lowers cost on the data warehouse.

Configure the CI job to run: `dbt build -s @state:modified --fail-fast`

* The `@state:modified` identifies the changed models

> * `--fail-fast` causes dbt to abort the first time an error is encountered to avoids wastefully running additional models when you already know there is a problem to resolve before merging the PR.

Reviewer: I might ask that we remove this section bc we're changing to defer to an environment in the next couple weeks. You'll still need a frequent dbt compile job in the prod environment, but I'd prefer to remove this section and write the best practices when the change is out. This way we don't risk a user need to change what they just did after following our instructions

Reviewer 4: @matt-winkler let's discuss this change request as ci improvements rollout is imminent.
